### ISSUE TITLE
bce(sigmoid(a), b) and bce_with_logits(a, b) return different values

### ISSUE THREAD

**Issue by kngwyu** [pytorch/pytorch#15655]
## 🐛 Bug

I'm not sure it's a bug or intended behavior, but
```
F.binary_cross_entropy(torch.sigmoid(a), b, reduction='sum')
```
and
```
F.binary_cross_entropy_with_logits(a, b, reduction='sum')
```
sometimes return different result.

## To Reproduce
This tar.gz file contains numpy arrays and a script to reproduce this issue.
[pytorch-issue15655.tar.gz](https://github.com/pytorch/pytorch/files/2720030/pytorch-issue15655.tar.gz)

1. `wget https://github.com/pytorch/pytorch/files/2720030/pytorch-issue15655.tar.gz`
2. `tar -xvf pytorch-issue15655.tar.gz`
3. `cd pytorch-issue15655`
4. `python3 repro.py`

`repro.py` contains
```python
import torch
from torch.nn import functional as F
import numpy as np
a, b = np.load('a.npy'), np.load('b.npy')
a, b = torch.tensor(a).cuda(), torch.tensor(b).cuda()
sce1 = F.binary_cross_entropy(torch.sigmoid(a), b, reduction='sum')
sce2 = F.binary_cross_entropy_with_logits(a, b, reduction='sum')
print('sce1: ', sce1)
print('sce2: ', sce2)
print('diff: ', sce1 - sce2)
```

## Expected behavior

I expect `diff == 0`

## Environment
PyTorch version: 1.0.0.dev20190101
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Arch Linux
GCC version: (GCC) 8.2.1 20181127
CMake version: version 3.13.1

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1050 Ti
Nvidia driver version: 415.25
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.7.4.1
/usr/lib/libcudnn_static_v7.a

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

**Comment by ngimel** [comment]
If you differences are small, this is likely different numerical stability of the two approaches, `binary_cross_entropy_with_logits` is more numerically stable https://pytorch.org/docs/stable/nn.html#torch.nn.BCEWithLogitsLoss

**Comment by kngwyu** [comment]
Thanks, but it doesn't look the problem of numerical stability.
I made a shorter example here.
```python
>>> import torch
>>> from torch.nn import functional as F
>>> a, b = torch.tensor([17.0, 15.0]), torch.zeros(2)
>>> F.binary_cross_entropy(torch.sigmoid(a), b, reduction='sum')
tensor(42.4748)
>>> F.binary_cross_entropy_with_logits(a, b, reduction='sum')
tensor(32.)
```

**Comment by ngimel** [comment]
Yes, that's the problem of numerical stability. Sigmoid of 17 and 15 is very close to 1, and loses precision. Convert a and b to double and see results become the same.

**Comment by kngwyu** [comment]
> Yes, that's the problem of numerical stability. Sigmoid of 17 and 15 is very close to 1, and loses precision. Convert a and b to double and see results become the same.

I got same results. Thanks.
